---
title: "Wildtype Analysis"
author: "Lachlan Baer"
date: "09/08/2019"
output: 
  html_document:
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# R Setup

```{r}
library(tidyverse)
library(magrittr)
library(edgeR)
library(AnnotationHub)
library(scales)
library(pander)
library(ggrepel)
library(ngsReports)
```

```{r}
setwd("~/Documents/Bioinformatics/Projects/2019_Lardelli_AllWT_6month_RNASeq/R")
```

# Metadata setup

```{r, include = FALSE}
oldName <- c("1_non_mutant_Q96_K97del_6mth_10_03_2016_S1_fem",
             "2_non_mutant_Q96_K97del_6mth_10_03_2016_S2_fem",
             "3_non_mutant_Q96_K97del_6mth_10_03_2016_S3_fem",
             "4_non_mutant_K97Gfs_6mth_10_03_2016_S1_fem",
             "5_non_mutant_K97Gfs_6mth_10_03_2016_S2_fem",
             "6_non_mutant_K97Gfs_6mth_10_03_2016_S3_fem",
             "9_Ps2Ex3M1_WT_6month_07_07_2016_F3_79_Fem",
             "10_Ps2Ex3M1_WT_6month_07_07_2016_F3_86_Fem",
             "11_Ps2Ex3M1_WT_6month_07_07_2016_F3_88_Fem",
             "12_Ps2Ex3M1_WT_6month_07_07_2016_F3_95_Fem",
             "1-MORGAN-6P-PN1_S2_R1_001",
             "2-MORGAN-6P-PN2_S5_R1_001",
             "3-MORGAN-6P-PN3_S10_R1_001",
             "4-MORGAN-6P-PN4_S8_R1_001",
             "11__-_5",
             "12__-_4",
             "13__-_3",
             "14__-_2",
             "15__-_1",
             "W1",
             "W2",
             "W3",
             "W4",
             "W5",
             "3_13W_1",
             "4_14W_1",
             "6_16W_1",
             "8_19W_1",
             "9_1W_1",
             "12_9W_1")
sample <- c("2016Q96K97_Wt1",
             "2016Q96K97_Wt2",
             "2016Q96K97_Wt3",
             "2016K97Gfs_Wt1",
             "2016K97Gfs_Wt2",
             "2016K97Gfs_Wt3",
             "2016PSEN2_Wt1",
             "2016PSEN2_Wt2",
             "2016PSEN2_Wt3",
             "2016PSEN2_Wt4",
             "2017Q96K97_Wt1",
             "2017Q96K97_Wt2",
             "2017Q96K97_Wt3",
             "2017Q96K97_Wt4",
             "2018PSEN2_Wt1",
             "2018PSEN2_Wt2",
             "2018PSEN2_Wt3",
             "2018PSEN2_Wt4",
             "2018PSEN2_Wt5",
             "2019Q96K97_Wt1",
             "2019Q96K97_Wt2",
             "2019Q96K97_Wt3",
             "2019Q96K97_Wt4",
             "2019Q96K97_Wt5",
             "2019SORL1_Wt1",
             "2019SORL1_Wt2",
             "2019SORL1_Wt3",
             "2019SORL1_Wt4",
             "2019SORL1_Wt5",
             "2019SORL1_Wt6")
sex <- c("F", "F", "F",
         "F", "F", "F",
         "F", "F", "F", "F",
         "M", "F", "F", "F",
         "F", "F", "F", "M", "M",
         "F", "F", "F", "F", "F",
         "F", "F", "M", "M", "M", "F")
platform <- c("Hiseq?", "Hiseq?", "Hiseq?",
              "Hiseq?", "Hiseq?", "Hiseq?",
              "Hiseq?", "Hiseq?", "Hiseq?", "Hiseq?",
              "Nextseq", "Nextseq", "Nextseq", "Nextseq",
              "Nextseq", "Nextseq", "Nextseq", "Nextseq", "Nextseq",
              "Novaseq", "Novaseq", "Novaseq", "Novaseq", "Novaseq",
              "Novaseq", "Novaseq", "Novaseq", "Novaseq", "Novaseq", "Novaseq")
provider <- c("ACRF", "ACRF", "ACRF",
              "ACRF", "ACRF", "ACRF",
              "ACRF", "ACRF", "ACRF", "ACRF",
              "SAHMRI", "SAHMRI", "SAHMRI", "SAHMRI",
              "SAHMRI", "SAHMRI", "SAHMRI", "SAHMRI", "SAHMRI",
              "Novogene", "Novogene", "Novogene", "Novogene", "Novogene",
              "Novogene", "Novogene", "Novogene", 
              "Novogene", "Novogene", "Novogene")
family <- c("2016Q96K97", "2016Q96K97", "2016Q96K97",
            "2016K97Gfs", "2016K97Gfs", "2016K97Gfs",
            "2016PSEN2", "2016PSEN2", "2016PSEN2", "2016PSEN2",
            "2017Q96K97", "2017Q96K97", "2017Q96K97", "2017Q96K97",
            "2018PSEN2", "2018PSEN2", "2018PSEN2", "2018PSEN2", "2018PSEN2",
            "2019Q96K97", "2019Q96K97", "2019Q96K97", "2019Q96K97", "2019Q96K97",
            "2019SORL1", "2019SORL1", "2019SORL1",
            "2019SORL1", "2019SORL1", "2019SORL1")
mutation <- c("Q96K97del", "Q96K97del", "Q96K97del",
              "K97Gfs", "K97Gfs", "K97Gfs",
              "S4Ter", "S4Ter", "S4Ter", "S4Ter",
              "Q96K97del", "Q96K97del", "Q96K97del", "Q96K97del",
              "FSvsFAD", "FSvsFAD", "FSvsFAD", "FSvsFAD", "FSvsFAD",
              "Q96K97del", "Q96K97del", "Q96K97del", "Q96K97del", "Q96K97del",
              "W1818X", "W1818X", "W1818X", "W1818X", "W1818X", "W1818X")
readLength <- c("150", "150", "150",
                "150", "100??", "150",
                "150", "150", "150", "150",
                "75", "75", "75", "75",
                "75", "75", "75", "75", "75",
                "150", "150", "150", "150", "150",
                "150", "150", "150", "150", "150", "150")
```

```{r}
metaData <- data.frame(
  sample, sex, family, mutation, provider, platform, readLength, oldName
  )
metaData
```

# ngsReports

```{r}
rawDataDir <- file.path("../0_rawData/FastQC/")
writeHtmlReport(rawDataDir)
```

```{r}
trimDataDir <- file.path("../1_trimmedData/FastQC/")
writeHtmlReport(trimDataDir)
```

```{r}
alignDataDir <- file.path("../2_alignedData/FastQC/")
writeHtmlReport(alignDataDir, overwrite = TRUE)
```

# PCA Analysis

## Load Data

```{r}
counts1 <- read_tsv("../2_alignedData/featureCounts/genes.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.),"Aligned.sortedByCoord.out.bam")) %>%
  set_colnames(str_remove(colnames(.),".fq.gz"))
counts2 <- read_tsv("../../20190122_Q96K97_NoStress_RNASeq/2_alignedData/featureCounts/genes_ss.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "_ssAligned.sortedByCoord.out.bam")) %>%
  dplyr::select(Geneid, starts_with("W"))
counts <- counts1 %>%
  full_join(counts2) %>% 
  column_to_rownames("Geneid") %>%
  set_colnames(metaData$sample[match(colnames(.), metaData$oldName)]) %>%
  rownames_to_column("Geneid")
## CHECK THIS WITH STEVE!
```

## Create DGE List

```{r}
dgeList <- counts %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  DGEList() %>%
  calcNormFactors()
```

```{r}
## Set group variables
dgeList$samples$group <- colnames(dgeList) %>%
  str_extract("201\\d(Q96K97|K97Gfs|PSEN2|SORL1)") %>%
  factor(
    levels = c(
      "2016Q96K97", "2016K97Gfs", "2016PSEN2", "2017Q96K97",
      "2018PSEN2", "2019Q96K97", "2019SORL1"
    )
  )
```

## Add gene information

```{r}
# Add AnnotationHub and subset to search for zebrafish
ah <- AnnotationHub()
ah %>%
  subset(species == "Danio rerio") %>%
  subset(dataprovider == "Ensembl") %>%
  subset(rdataclass == "EnsDb")
```

```{r}
# Select correct Ensembl release
ensDb <- ah[["AH69169"]]
```

```{r}
# Extract GenomicRanges object from ensDb
genesGR <- genes(ensDb)
```

```{r}
# Remove redundant columns from mcols
mcols(genesGR) <- mcols(genesGR)[c("gene_id", "gene_name", 
                                   "gene_biotype", "entrezid")]
```

```{r}
# Add genesGR to DGEList using rownames of DGEList to reorder the genesGR
dgeList$genes <- genesGR[rownames(dgeList),]
```

## Data QC

```{r}
# Perform logical test to see how many genes were not detected in dataset
dgeList$counts %>%
  rowSums() %>%
  is_greater_than(0) %>%
  table()
```

```{r}
# Check for genes having >= 3 samples with cpm > 1
dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(3) %>%
  table()
```

```{r}
# Create logical vector of genes to keep that fit criteria
genes2keep <- dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(3)
```

```{r}
# Create new DGEList of genes fitting criteria
dgeFilt <- dgeList[genes2keep,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r fig.align = "center"}
# Compare distributions of the DGELists before and after filtering
par(mfrow = c(1,2))
dgeList %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "Before Filtering")
dgeFilt %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "After Filtering")
par(mfrow = c(1,1))
```

## Library sizes

```{r fig.align = "center"}
# Check library sizes with box plot
dgeFilt$samples %>%
  ggplot(aes(group, lib.size, fill = group)) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  labs(x = "Family", y = "Library Size") +
  theme_bw() +
  theme(legend.position = "none")
```

## PCA

```{r}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
pca <- dgeFilt %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(pca)$importance %>% pander(split.tables = Inf)
```

```{r fig.align = "center"}
# Plot PCA
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(dgeFilt$samples, "sample")) %>%
  left_join(metaData) %>%
  ggplot(aes(PC1, PC2, colour = group, shape = provider, label = sex)) +
  geom_point(size = 3, stroke = 0.5) +
  # geom_text_repel() +
  labs(colour = "Family", shape = "Provider") +
  xlab(label = paste0("PC1 (", percent(summary(pca)$importance[2]), ")")) +
  ylab(label = paste0("PC2 (", percent(summary(pca)$importance[5]), ")")) +
  theme_bw()
```