---
title: "Wildtype Analysis"
author: "Lachlan Baer"
date: "09/08/2019"
output: 
  html_document:
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# **R Setup**

```{r}
library(tidyverse)
library(magrittr)
library(edgeR)
library(AnnotationHub)
library(scales)
library(pander)
library(ggrepel)
library(grid)
library(EDASeq)
library(multcomp)
library(org.Dr.eg.db)
library(org.Hs.eg.db)
```

```{r options}
theme_set(theme_bw())
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
panderOptions("table.style", "rmarkdown")
if (interactive()) setwd(here::here("R"))
```

# **Metadata setup**

```{r, include = FALSE}
oldName <- c("1_non_mutant_Q96_K97del_6mth_10_03_2016_S1_fem",
             "2_non_mutant_Q96_K97del_6mth_10_03_2016_S2_fem",
             "3_non_mutant_Q96_K97del_6mth_10_03_2016_S3_fem",
             "4_non_mutant_K97Gfs_6mth_10_03_2016_S1_fem",
             "5_non_mutant_K97Gfs_6mth_10_03_2016_S2_fem",
             "6_non_mutant_K97Gfs_6mth_10_03_2016_S3_fem",
             "9_Ps2Ex3M1_WT_6month_07_07_2016_F3_79_Fem",
             "10_Ps2Ex3M1_WT_6month_07_07_2016_F3_86_Fem",
             "11_Ps2Ex3M1_WT_6month_07_07_2016_F3_88_Fem",
             "12_Ps2Ex3M1_WT_6month_07_07_2016_F3_95_Fem",
             "1-MORGAN-6P-PN1_S2_R1_001",
             "2-MORGAN-6P-PN2_S5_R1_001",
             "3-MORGAN-6P-PN3_S10_R1_001",
             "4-MORGAN-6P-PN4_S8_R1_001",
             "11__-_5",
             "12__-_4",
             "13__-_3",
             "14__-_2",
             "15__-_1",
             "W1",
             "W2",
             "W3",
             "W4",
             "W5",
             "3_13W_1",
             "4_14W_1",
             "6_16W_1",
             "8_19W_1",
             "9_1W_1",
             "12_9W_1")
sample <- c("2016Q96K97_Wt1",
             "2016Q96K97_Wt2",
             "2016Q96K97_Wt3",
             "2016K97Gfs_Wt1",
             "2016K97Gfs_Wt2",
             "2016K97Gfs_Wt3",
             "2016PSEN2_Wt1",
             "2016PSEN2_Wt2",
             "2016PSEN2_Wt3",
             "2016PSEN2_Wt4",
             "2017Q96K97_Wt1",
             "2017Q96K97_Wt2",
             "2017Q96K97_Wt3",
             "2017Q96K97_Wt4",
             "2018PSEN2_Wt1",
             "2018PSEN2_Wt2",
             "2018PSEN2_Wt3",
             "2018PSEN2_Wt4",
             "2018PSEN2_Wt5",
             "2019Q96K97_Wt1",
             "2019Q96K97_Wt2",
             "2019Q96K97_Wt3",
             "2019Q96K97_Wt4",
             "2019Q96K97_Wt5",
             "2019SORL1_Wt1",
             "2019SORL1_Wt2",
             "2019SORL1_Wt3",
             "2019SORL1_Wt4",
             "2019SORL1_Wt5",
             "2019SORL1_Wt6")
sex <- c("F", "F", "F",
         "F", "F", "F",
         "F", "F", "F", "F",
         "M", "F", "F", "F",
         "F", "F", "F", "M", "M",
         "F", "F", "F", "F", "F",
         "F", "F", "M", "M", "M", "F")
platform <- c("Hiseq?", "Hiseq?", "Hiseq?",
              "Hiseq?", "Hiseq?", "Hiseq?",
              "Hiseq?", "Hiseq?", "Hiseq?", "Hiseq?",
              "Nextseq", "Nextseq", "Nextseq", "Nextseq",
              "Nextseq", "Nextseq", "Nextseq", "Nextseq", "Nextseq",
              "Novaseq", "Novaseq", "Novaseq", "Novaseq", "Novaseq",
              "Novaseq", "Novaseq", "Novaseq", "Novaseq", "Novaseq", "Novaseq")
provider <- c("ACRF", "ACRF", "ACRF",
              "ACRF", "ACRF", "ACRF",
              "ACRF", "ACRF", "ACRF", "ACRF",
              "SAHMRI", "SAHMRI", "SAHMRI", "SAHMRI",
              "SAHMRI", "SAHMRI", "SAHMRI", "SAHMRI", "SAHMRI",
              "Novogene", "Novogene", "Novogene", "Novogene", "Novogene",
              "Novogene", "Novogene", "Novogene", 
              "Novogene", "Novogene", "Novogene")
family <- c("2016Q96K97", "2016Q96K97", "2016Q96K97",
            "2016K97Gfs", "2016K97Gfs", "2016K97Gfs",
            "2016PSEN2", "2016PSEN2", "2016PSEN2", "2016PSEN2",
            "2017Q96K97", "2017Q96K97", "2017Q96K97", "2017Q96K97",
            "2018PSEN2", "2018PSEN2", "2018PSEN2", "2018PSEN2", "2018PSEN2",
            "2019Q96K97", "2019Q96K97", "2019Q96K97", "2019Q96K97", "2019Q96K97",
            "2019SORL1", "2019SORL1", "2019SORL1",
            "2019SORL1", "2019SORL1", "2019SORL1")
mutation <- c("Q96K97del", "Q96K97del", "Q96K97del",
              "K97Gfs", "K97Gfs", "K97Gfs",
              "S4Ter", "S4Ter", "S4Ter", "S4Ter",
              "Q96K97del", "Q96K97del", "Q96K97del", "Q96K97del",
              "FSvsFAD", "FSvsFAD", "FSvsFAD", "FSvsFAD", "FSvsFAD",
              "Q96K97del", "Q96K97del", "Q96K97del", "Q96K97del", "Q96K97del",
              "W1818X", "W1818X", "W1818X", "W1818X", "W1818X", "W1818X")
readLength <- c("150", "150", "150",
                "150", "100??", "150",
                "150", "150", "150", "150",
                "75", "75", "75", "75",
                "75", "75", "75", "75", "75",
                "150", "150", "150", "150", "150",
                "150", "150", "150", "150", "150", "150")
```

```{r}
metaData <- data.frame(
  sample, sex, family, mutation, provider, platform, readLength, oldName
  )
metaData %>% pander()
```

# **Data setup**

## Load data

```{r}
counts1 <- read_tsv("../2_alignedData/featureCounts/genes.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.),"Aligned.sortedByCoord.out.bam")) %>%
  set_colnames(str_remove(colnames(.),".fq.gz"))
counts2 <- read_tsv("../../20190122_Q96K97_NoStress_RNASeq/2_alignedData/featureCounts/genes_ss.out") %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "_ssAligned.sortedByCoord.out.bam")) %>%
  dplyr::select(Geneid, starts_with("W"))
counts <- counts1 %>%
  full_join(counts2) %>% 
  column_to_rownames("Geneid") %>%
  set_colnames(metaData$sample[match(colnames(.), metaData$oldName)]) %>%
  rownames_to_column("Geneid")
```

## Create DGE List

```{r}
dgeList <- counts %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  DGEList() %>%
  calcNormFactors()
```

```{r}
# Set group variables
dgeList$samples$group <- colnames(dgeList) %>%
  str_extract("201\\d(Q96K97|K97Gfs|PSEN2|SORL1)") %>%
  factor(
    levels = c(
      "2016Q96K97", "2016K97Gfs", "2016PSEN2", "2017Q96K97",
      "2018PSEN2", "2019Q96K97", "2019SORL1"
    )
  )
dgeList$samples$sample <- colnames(dgeList)
```

## Add gene information

```{r}
# Add AnnotationHub and subset to search for zebrafish
ah <- AnnotationHub()
ah %>%
  subset(species == "Danio rerio") %>%
  subset(dataprovider == "Ensembl") %>%
  subset(rdataclass == "EnsDb")
```

```{r}
# Select correct Ensembl release
ensDb <- ah[["AH69169"]]
```

```{r}
# Extract GenomicRanges object from ensDb
genesGR <- genes(ensDb)
```

```{r}
# Remove redundant columns from mcols
mcols(genesGR) <- mcols(genesGR)[c("gene_id", "gene_name", 
                                   "gene_biotype", "entrezid")]
```

```{r}
# Add genesGR to DGEList using rownames of DGEList to reorder the genesGR
dgeList$genes <- genesGR[rownames(dgeList),]
```

## Data QC

```{r}
# Perform logical test to see how many genes were not detected in dataset
dgeList$counts %>%
  rowSums() %>%
  is_greater_than(0) %>%
  table()
```

```{r}
# Check for genes having >= 3 samples with cpm > 1
dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(3) %>%
  table()
```

```{r}
# Create logical vector of genes to keep that fit criteria
genes2keep <- dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(3)
```

```{r}
# Create new DGEList of genes fitting criteria
dgeFilt <- dgeList[genes2keep,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r fig.align = "center"}
# Compare distributions of the DGELists before and after filtering
par(mfrow = c(1,2))
dgeList %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "Before Filtering")
dgeFilt %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "After Filtering")
par(mfrow = c(1,1))
```

## Library size check

```{r fig.align = "center"}
# Check library sizes with box plot
dgeFilt$samples %>%
  ggplot(aes(group, lib.size, fill = group)) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  labs(x = "Family", y = "Library Size") +
  theme(legend.position = "none")
```

# **PCA**

## Setup

```{r}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
pca <- dgeFilt %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(pca)$importance %>% pander()
```

```{r fig.align = "center"}
# Create function for plotting PCA
ggPCA <- function(x, y){
  pca$x %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    as_tibble() %>%
    dplyr::select(sample, one_of(c(x, y))) %>%
    left_join(dgeFilt$samples) %>%
    left_join(metaData) %>%
    ggplot(
      aes_string(
        x, y, colour = "group", shape = "provider", label = "sex"
      )
    ) +
    geom_point(size = 3, stroke = 0.5) +
    geom_text_repel(show.legend = FALSE) +
    # labs(colour = "Family", shape = "Provider") +
    xlab(label = paste0(x, " (", percent(summary(pca)$importance[2, x]), ")")) +
    ylab(label = paste0(y, " (", percent(summary(pca)$importance[2, y]), ")")) +
    theme_bw()
}
```

## Plots

```{r}
ggPCA("PC1", "PC2")
```

```{r}
# Set up grid for visualisation of 3 PCA's at once.
vp1 <- viewport(x = 0.25, y = 0.75, width = 0.5, height = 0.5)
vp2 <- viewport(x = 0.75, y = 0.75, width = 0.5, height = 0.5)
vp3 <- viewport(x = 0.25, y = 0.25, width = 0.5, height = 0.5)
# Plot PCA plots in grid
if (interactive()) grid::grid.newpage()
print(
  ggPCA("PC2", "PC3") +
    theme(legend.position = "none"),
  vp = vp1
)
print(
  ggPCA("PC3", "PC4") +
    theme(legend.position = "none"),
  vp = vp2
)
print(
  ggPCA("PC4", "PC5") +
    guides(
      colour = guide_legend(ncol = 2),
      shape = guide_legend(ncol = 3)
    ) +
    theme(legend.position = c(1.6, 0.5)),
  vp = vp3
)
```

# **Differential Expression**

## Gene labeller

```{r}
# Set up gene labeller for easy conversion
geneLabeller <- as_labeller(
  structure(genesGR$gene_name, names = genesGR$gene_id)
)
```

## Voom

```{r}
voomData <- voom(dgeFilt, design = matrix(1, nrow = ncol(dgeFilt)))
```

```{r}
pcaDesign <- cbind(Intercept = 1, pca$x[,1:6])
```

```{r}
pcaFit <- lmFit(voomData, pcaDesign) %>%
  eBayes
```

```{r}
# topTable of PC1
topTablePC1 <- topTable(pcaFit, coef = "PC1", n = Inf) %>%
  as_tibble() %>% 
  set_colnames(str_remove_all(colnames(.), "ID.")) %>%
  unite("Range", start, end, sep = "-") %>%
  unite("Location", seqnames, Range, strand, sep = ":") %>%
  dplyr::select(
    gene_id, gene_name, AveExpr, logFC, 
    P.Value, adj.P.Val, t, Location, entrezid
  )
```

Just grab one or two to test:

```{r}
voomData$E[rownames(topTable(pcaFit, coef = "PC1", n = Inf))[1:10],] %>%
  t() %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  gather("gene_id", "logCPM", -sample) %>%
  left_join(voomData$targets) %>%
  left_join(metaData, by = "sample") %>%
  cbind(pca$x[.$sample,]) %>%
  ggplot(aes(PC1, logCPM, colour = group, shape = provider, label = sex)) +
  geom_point() +
  facet_wrap(~gene_id, labeller = geneLabeller)
```

# **Gene length and GC content**

## Setup

```{r}
# # This is extremely slow, takes approximately 30 minutes.
# # Saved as .rds file to save time
# id <- rownames(dgeFilt)
# gcContent <- getGeneLengthAndGCContent(id, org = "dre", mode = "biomart")
# saveRDS(gcContent, "../files/gcContent.rds")
```

```{r}
geneGCLen <- readRDS("../files/gcContent.rds") %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  as_tibble()
```

```{r}
# Get transcript lengths and choose longest transcript to be gene length
transLen <- transcriptLengths(ensDb) %>% as_tibble()
geneLen <- transLen %>%
  arrange(gene_id, desc(tx_len)) %>%
  distinct(gene_id, .keep_all = TRUE)
```

## Gene Length

### Initial inspection

```{r}
topTablePC1 %>% 
  mutate(rank = seq_len(nrow(.))) %>% 
  left_join(geneGCLen) %>% 
  ggplot(aes(rank, length)) + 
  geom_point() + 
  scale_y_log10() +
  geom_smooth(method = "lm", se = FALSE)
```

```{r}
topTablePC1 %>%
  mutate(rank = seq_len(nrow(.))) %>% 
  left_join(geneGCLen) %>%
  ggplot(aes(t, length)) +
  geom_point() +
  scale_y_log10() +
  geom_smooth(method = "lm", se = FALSE)
```

### Length comparison

```{r}
# Grab DE down genes and join lengths
# Outliers were also removed to make boxplots look nicer (only 7 genes)
deDown <- topTablePC1 %>% 
  dplyr::filter(adj.P.Val < 0.05, logFC < 0) %>%
  dplyr::select(gene_id, gene_name) %>%
  dplyr::mutate(group = "down")
```

```{r}
# Grab DE up genes and join lengths
deUp <- topTablePC1 %>% 
  dplyr::filter(adj.P.Val < 0.05, logFC > 0) %>%
  dplyr::select(gene_id, gene_name) %>%
  dplyr::mutate(group = "up")
```

```{r}
# Grab non DE genes and join lengths
deNon <- topTablePC1 %>% 
  dplyr::filter(adj.P.Val > 0.05) %>%
  dplyr::select(gene_id, gene_name) %>%
  dplyr::mutate(group = "non")
```

```{r}
# Join into one tibble and box plot to compare
deDown %>%
  full_join(deUp) %>%
  full_join(deNon) %>%
  left_join(geneGCLen) %>%
  ggplot(aes(group, length, fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  scale_y_continuous(labels = comma) +
  labs(x = "DE", y = "Gene Length") +
  coord_cartesian(ylim = c(0, 10000)) +
  theme(legend.position = "none") 
```

### LogFC vs length

```{r}
topTablePC1 %>%
  left_join(geneGCLen) %>%
  mutate(logFC = logFC^(1/3))
  ggplot(aes(length, logFC, colour = gc)) +
  geom_point(alpha = 0.5) +
  scale_color_gradient2(low = "black", mid = "red", high = "yellow") +
  scale_x_log10(labels = comma) +
  xlab("Length (bp)") +
  ylab("LogFC")
```

## GC content

### GC content comparison

```{r}
deDown %>%
  full_join(deUp) %>%
  full_join(deNon) %>% 
  left_join(geneGCLen) %>%
  ggplot(aes(group, gc, fill = group)) +
  geom_boxplot() +
  labs(x = "DE", y = "GC Content") +
  theme(legend.position = "none")
```

### LogFC vs GC content

```{r}
topTablePC1 %>%
  left_join(geneGCLen) %>%
  ggplot(aes(gc, logFC)) +
  geom_point() +
  scale_x_continuous(label = scales::percent_format(accuracy = 1)) +
  xlab("GC content") +
  ylab("LogFC")
```

# **Gene level variance**

```{r}
# Create function for isolating columns per family and calculating variance
famVar <- function(x){
  dgeFilt %>%
    cpm() %>%
    as.data.frame() %>%
    dplyr::select(str_subset(colnames(.), paste0(x, "_Wt\\d"))) %>%
    apply(1, var) %>%
    as_tibble(rownames = "geneid") %>%
    dplyr::mutate(group = x)
}
```

```{r}
# Calculate variance for each family
var1 <- famVar("2016Q96K97")
var2 <- famVar("2016K97Gfs") 
var3 <- famVar("2016PSEN2")
var4 <- famVar("2017Q96K97")
var5 <- famVar("2018PSEN2")
var6 <- famVar("2019Q96K97")
var7 <- famVar("2019SORL1")
```

```{r}
# Plot boxplot
var1 %>% 
  full_join(var2) %>%
  full_join(var3) %>%
  full_join(var4) %>%
  full_join(var5) %>%
  full_join(var6) %>%
  full_join(var7) %>%
  ggplot(aes(group, value, fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  labs(x = "Family", y = "Gene level variance", fill = "Provider") +
  scale_fill_discrete(
    labels = c(
      "ACRF", "ACRF", "ACRF", "SAHMRI", "SAHMRI", "Novogene", "Novogene"
    )
  ) +
  coord_cartesian(ylim = c(0, 75))
```

# **GO analysis**

## ID conversion

```{r}
# Some packages are required to be unloaded before biomaRt can be loaded
detach("package:EDASeq", unload = TRUE)
detach("package:ensembldb", unload = TRUE)
detach("package:GenomicFeatures", unload = TRUE)
library(biomaRt)
```

```{r}
# Human entrez and ensembl IDs
humanEntrezEns <- org.Hs.egENSEMBL %>%
  as.data.frame %>%
  set_colnames(c("Hs_entrez", "Hs_ensembl"))
```

```{r}
# Zebrafish ensembl IDs
zebEns <- org.Dr.egENSEMBL %>%
  as.data.frame %>%
  set_colnames(c("Dr_entrez", "Dr_ensembl"))
```

```{r}
# Create a data.frame to map between human entrezgenes and zebrafish ensembl IDs.
# BioMart only includes homolog mappings for ensembl IDs which is why we need to 
# retrieve human ensembl IDs, then join to the humanEntrezEns data.frame,
# in order to get the desired human entrezgenes to zebrafish ensembl ID mapping. 
zebMart <- useMart("ENSEMBL_MART_ENSEMBL", "drerio_gene_ensembl")
getFromBiomart <- c("ensembl_gene_id", "hsapiens_homolog_ensembl_gene")
ens2entrez <- getBM(
  getFromBiomart,
  values = unique(zebEns$Dr_ensembl),
  mart = zebMart
) %>%
  set_colnames(c("Dr_ensembl", "Hs_ensembl")) %>%
  left_join(humanEntrezEns, by = "Hs_ensembl") %>%
  dplyr::select(-Hs_ensembl) %>% 
  dplyr::filter(complete.cases(.)) %>%
  as_tibble()
```

## GO enrichment

```{r}
de <- topTablePC1 %>%
  dplyr::filter(adj.P.Val < 0.05) %>%
  dplyr::select(gene_id) %>%
  left_join(ens2entrez, by = c("gene_id" = "Dr_ensembl")) %>%
  dplyr::filter(!is.na(Hs_entrez)) %>%
  .[["Hs_entrez"]] %>%
  unique()
uv <- topTablePC1 %>%
  dplyr::select(gene_id) %>%
  left_join(ens2entrez, by = c("gene_id" = "Dr_ensembl")) %>%
  dplyr::filter(!is.na(Hs_entrez)) %>%
  .[["Hs_entrez"]] %>%
  unique()
```

```{r}
goResults <- goana(de = de, universe = uv, species = "Hs")
```

```{r}
goResults %>% 
  rownames_to_column("GO ID") %>%
  as_tibble() %>%
  dplyr::filter(DE > 1) %>%
  dplyr::arrange(P.DE) %>%
  mutate(FDR = p.adjust(P.DE, "fdr")) %>%
  dplyr::filter(FDR < 0.05) %>%
  mutate(`GO ID` = str_replace(`GO ID`, ":", "\\\\:")) %>%
  pander(justify = "llrrrrr")
```

# Steve's wizardry

```{r}
myLmPlot <- function(x){
  par(mfrow = c(2,2))
  plot(x)
  par(mfrow = c(1,1))
}
```

We ran an initial model plotting gene-level contributions (i.e. rotations) to PC1 against the two potential predictor variables: `log10(length)` and `gc` content.

```{r}
geneGCLen %>%
  cbind(pca$rotation[.$gene_id,]) %>% 
  as_tibble() %>%
  lm(PC1 ~ log10(length)*gc, data = .) %T>%
  myLmPlot() %>%
  summary() %>%
  pander()
```

This model clearly violated the assumption of constant variance meaning the basic assumptions for the linear model were not able to be applied.
As such, the Box-Cox transformation was applied using an offset of $\gamma = 1$ and with $\lambda = 38$, as determined by fitting the profile likelihood for $\lambda$.

```{r}
geneGCLen %>%
  cbind(pca$rotation[.$gene_id,]) %>% 
  as_tibble() %>%
  with(
    boxCox((1 + PC1) ~ log(length)*gc,
           lambda = seq(0, 50, length = 50)
           )
    )
```

The transformed data was then fit and the assumption of constant variance was able to be applied.

```{r}
geneGCLen %>%
  cbind(pca$rotation[.$gene_id,]) %>% 
  as_tibble() %>%
  mutate(
    bcPC1 = bcnPower(PC1, lambda = 38, gamma = 1)
  ) %>%
  lm(bcPC1 ~ log10(length)*gc, data = .) %T>%
  myLmPlot() %>%
  summary() %>%
  pander()
```

Whilst not being simply interpretable, this suggests that both GC content and gene length are contributing to the variability detected along PC1.

```{r}
geneGCLen %>%
  cbind(pca$rotation[.$gene_id,]) %>% 
  as_tibble() %>%
  ggplot(aes(length, PC1, colour = gc)) +
  geom_point() +
  scale_x_log10(labels = comma) +
  scale_colour_viridis_c(option = "inferno")
```

--------------------------------------------------------------------------------

## Current method
  
```{r}
nBins <- list(length = 10, gc = 10)
bin_lm <- geneGCLen %>%
  cbind(pca$rotation[.$gene_id,]) %>% 
  as_tibble() %>%
  mutate(
    lengthBins = cut(
      log(length), 
      breaks = nBins$length, 
      labels = paste0("L", seq_len(nBins$length)), 
      include.lowest = TRUE
    ),
    gcBins = cut(
      gc, 
      breaks = nBins$gc, 
      labels = paste0("GC", seq_len(nBins$gc)), 
      include.lowest = TRUE
    ),
    bothBins = paste(lengthBins, gcBins, sep = "_"),
    bothBins = as.factor(bothBins)
  ) %>%
  group_by(bothBins) %>%
  mutate(n = n()) %>%
  ungroup() %>%
  dplyr::filter(n > 1) %>%
  lm(PC1 ~ 0 + bothBins, data = .)
```

```{r}
# To use different significance adjustment
binFDR <- bin_lm %>% 
  glht() %>% 
  summary(test = adjusted("fdr"))
```

```{r}
# Extract p-values 
binSig <- binFDR$test$pvalues %>% enframe() %>% 
  mutate(name = str_remove(name, "bothBins")) %>% 
  separate(name, into = c("length", "GC")) %>%
  dplyr::rename(p = value) %>%
  dplyr::mutate(sig = case_when(
    between(p, 0, 0.001) ~ 5,
    between(p, 0.001, 0.01) ~ 4,
    between(p, 0.01, 0.05) ~ 3,
    between(p, 0.05, 0.1) ~ 2,
    between(p, 0.1, 1) ~ 1
  )) %>%
  dplyr::select(length, GC, sig)
```

```{r}
coef(bin_lm) %>% 
  enframe() %>%
  mutate(name = str_remove(name, "bothBins")) %>% 
  separate(name, into = c("length", "GC")) %>%
  left_join(binSig) %>%
  mutate(
    length = str_remove(length, "L") %>% as.integer(),
    GC = str_remove(GC, "GC") %>% as.integer()
  ) %>%
  ggplot(aes(length, GC)) +
  geom_point(aes(colour = value, size = sig)) +
  scale_size_continuous(labels = c(
    "0 < p < 0.001",
    "0.001 < p < 0.01",
    "0.01 < p < 0.1",
    "0.5 < p < 0.1",
    "0.1 < p < 1"
    )) +
  scale_x_continuous(breaks = seq_len(nBins$length)) + 
  scale_y_continuous(breaks = seq_len(nBins$gc)) +
  labs(
    x = "Length Bin",
    y = "GC Content Bin",
    colour = paste("Contribution", "to PC1", sep = "\n"),
    size = "Significance"
  ) +
  theme(
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank()
  )
```

--------------------------------------------------------------------------------

```{r}
# bin_conf <- glht(bin_lm, linfct = mcp(bothBins = "Tukey")) %>%
#   confint() %>%
#   .[["confint"]] %>%
#   as.data.frame() %>%
#   rownames_to_column("Comparison") %>%
#   as_tibble() %>%
#   mutate(Sig = case_when(
#     sign(lwr) == sign(upr) ~ sign(lwr),
#     sign(lwr) != sign(upr) ~ 0
#   ))
```

```{r}
# bin_conf %>%
#   ggplot(aes(Estimate, Comparison, colour = Sig)) +
#   geom_point() +
#   geom_errorbarh(aes(xmin = lwr, xmax = upr)) +
#   geom_vline(xintercept = 0) +
#   scale_colour_gradient2(low = "red", mid = "black", high = "red") +
#   theme(
#     axis.text.y = element_text(size = 5)
#   )
```

```{r}
# list(
#   bin_conf %>% 
#     separate(Comparison, c("x", "y"), " - "), 
#   bin_conf %>% 
#     separate(Comparison, c("y", "x"), " - ") %>% 
#     mutate(Sig = -1*Sig)
# ) %>%
#   bind_rows() %>%
#   mutate(
#     L1x = grepl("L1", x),
#     L2x = grepl("L2", x),
#     L3x = grepl("L3", x),
#     L4x = grepl("L4", x),
#     L5x = grepl("L5", x)
#   ) %>%
#   ggplot(aes(x, y, fill = Sig)) +
#   scale_fill_gradient2(low = "darkred", mid = "white", high = "darkgreen") +
#   geom_raster() + 
#   xlab("") +
#   ylab("") +
#   facet_wrap(
#     ~L5x + L4x + L3x + L2x + L1x, 
#     scales = "free"
#   ) +
#   theme(
#     strip.background = element_blank(),
#     strip.text.x = element_blank(),
#     strip.text.y = element_blank(),
#     axis.text.x = element_text(angle = -55, hjust = -0.1),
#     panel.grid.major = element_blank(), 
#     panel.grid.minor = element_blank()
#   )
```
